File: src/services/parser/v2/group.ts
import { config } from "../../../../config";
import { WeekIndex, getShortSubjectName, mergeDays } from "../../../utils";
import { AbstractParser } from "../abstract";
import { GroupDay, GroupLesson, GroupLessonExplain, Groups } from "../types/group";
import { buildTableGrid, findHeaderRowIndex, getDayRangesFromGrid } from "./grid";
import { DayRange, GridCell } from "./types";
import { cleanText, extractLines, isTeacherLine, parseLessonNumber } from "./text";

type ParsedTable = {
    group: string,
    groupNumber: string,
    days: GroupDay[]
}

export default class StudentParserV2 extends AbstractParser {
    protected groups: Groups = {};

    public run(): Groups {
        const tables = this.findTables();

        for (const table of tables) {
            const label = this.findHeading(table, 'группа');
            if (!label) {
                continue;
            }

            const parsed = this.parseTable(table, label);
            if (!parsed) {
                continue;
            }

            const existing = this.groups[parsed.groupNumber];
            if (existing) {
                if (config.parser.v2?.allowTwoTables === false) {
                    continue;
                }
                const merged = mergeDays(parsed.days, existing.days);
                existing.days = merged.mergedDays;
                existing.group = parsed.group;
            } else {
                this.groups[parsed.groupNumber] = {
                    group: parsed.group,
                    days: parsed.days
                };
            }
        }

        this.clearSundays(this.groups);

        for (const group in this.groups) {
            for (const day of this.groups[group].days) {
                this.postProcessDay(day);
            }
        }

        return this.groups;
    }

    protected parseTable(table: HTMLTableElement, label: string): ParsedTable | null {
        const group = this.parseGroupLabel(label);
        const groupNumber = group ? this.parseGroupNumber(group) : undefined;
        if (!group || !groupNumber) {
            return null;
        }

        const { grid } = buildTableGrid(table);
        const minDays = config.parser.v2?.minDaysInTable ?? 5;
        const headerRowIndex = findHeaderRowIndex(grid, config.parser.v2?.headerScanRows ?? 5, minDays);
        if (headerRowIndex === null) {
            return null;
        }
        const ranges = this.getDayRanges(grid, headerRowIndex);
        if (!ranges.length) {
            return null;
        }

        const days: GroupDay[] = ranges.map((range) => ({
            day: range.day,
            lessons: []
        }));

        for (let rowIndex = headerRowIndex + 1; rowIndex < grid.length; rowIndex++) {
            const row = grid[rowIndex];
            if (!row) {
                continue;
            }

            const numberCell = row[0];
            if (!numberCell || numberCell.row !== rowIndex) {
                continue;
            }

            const lessonNumber = parseLessonNumber(numberCell.cell.textContent);
            if (!lessonNumber) {
                continue;
            }

            for (let i = 0; i < ranges.length; i++) {
                const range = ranges[i];
                const lessonCell = this.getCell(row, rowIndex, range.start);
                const cabinetCell = range.span > 1 ? this.getCell(row, rowIndex, range.start + 1) : undefined;

                const lesson = this.parseLesson(lessonCell, cabinetCell);
                this.assignLesson(days[i], lessonNumber, lesson);
            }
        }

        return {
            group,
            groupNumber,
            days
        };
    }

    protected parseGroupLabel(label: string): string | null {
        const text = cleanText(label);
        if (!text) {
            return null;
        }

        if (!text.toLowerCase().startsWith('группа')) {
            return null;
        }

        const parts = text.split('-');
        if (parts.length < 2) {
            return null;
        }

        return parts.slice(1).join('-').trim();
    }

    protected findTables(): HTMLTableElement[] {
        const tables = Array.from(this.content.querySelectorAll('table') as NodeListOf<HTMLTableElement>);
        const candidates: { table: HTMLTableElement, weekIndex: number }[] = [];
        const minDays = config.parser.v2?.minDaysInTable ?? 5;
        const headerScanRows = config.parser.v2?.headerScanRows ?? 5;

        for (const table of tables) {
            const { grid } = buildTableGrid(table);
            const headerRowIndex = findHeaderRowIndex(grid, headerScanRows, minDays);
            if (headerRowIndex === null) {
                continue;
            }

            const ranges = this.getDayRanges(grid, headerRowIndex);
            if (!ranges.length) {
                continue;
            }

            const weekIndexes = new Set<number>();
            for (const range of ranges) {
                try {
                    weekIndexes.add(WeekIndex.fromStringDate(range.day).valueOf());
                } catch {
                    continue;
                }
            }

            if (weekIndexes.size === 0) {
                continue;
            }

            const first = weekIndexes.values().next().value;
            if (typeof first !== 'number') {
                continue;
            }

            candidates.push({
                table,
                weekIndex: first
            });
        }

        if (!candidates.length) {
            return [];
        }

        const currentWeek = WeekIndex.now().valueOf();
        const holdCurrentOnSunday = config.parser.v2?.sundayHoldCurrent ?? true;
        const isSunday = new Date().getDay() === 0;
        const weekPolicy = config.parser.v2?.weekPolicy ?? 'preferCurrent';
        const weekIndexes = Array.from(new Set(candidates.map((item) => item.weekIndex)));
        let targetWeek = weekIndexes[0];

        if (weekIndexes.includes(currentWeek)) {
            targetWeek = currentWeek;
        } else if (weekPolicy === 'closest') {
            targetWeek = weekIndexes.sort((a, b) => Math.abs(a - currentWeek) - Math.abs(b - currentWeek))[0];
        } else if (weekPolicy === 'preferCurrent') {
            const future = weekIndexes.filter((value) => value > currentWeek).sort((a, b) => a - b);
            const past = weekIndexes.filter((value) => value < currentWeek).sort((a, b) => b - a);
            if (holdCurrentOnSunday && isSunday && past.length) {
                targetWeek = past[0];
            } else {
                targetWeek = future.length ? future[0] : weekIndexes.sort((a, b) => Math.abs(a - currentWeek) - Math.abs(b - currentWeek))[0];
            }
        }

        return candidates.filter((item) => item.weekIndex === targetWeek).map((item) => item.table);
    }

    protected findHeading(table: HTMLTableElement, keyword: string): string | null {
        let element = table.previousElementSibling;
        let depth = 0;

        while (element && depth < 12) {
            const tag = element.tagName.toLowerCase();
            if (['h1', 'h2', 'h3', 'h4'].includes(tag)) {
                const text = cleanText(element.textContent);
                if (text && text.toLowerCase().includes(keyword)) {
                    return text;
                }
            }

            if (tag === 'table') {
                break;
            }

            element = element.previousElementSibling;
            depth += 1;
        }

        return null;
    }

    protected getDayRanges(grid: GridCell[][], headerRowIndex: number): DayRange[] {
        const ranges = getDayRangesFromGrid(grid, headerRowIndex);
        const minDays = config.parser.v2?.minDaysInTable ?? 5;
        return ranges.length >= minDays ? ranges : [];
    }

    protected assignLesson(day: GroupDay, lessonNumber: number, lesson: GroupLesson) {
        const index = lessonNumber - 1;

        while (day.lessons.length < index) {
            day.lessons.push(null);
        }

        if (day.lessons.length === index) {
            day.lessons.push(lesson);
        } else {
            day.lessons[index] = lesson;
        }
    }

    protected parseLesson(lessonCell?: HTMLTableCellElement | null, cabinetCell?: HTMLTableCellElement | null): GroupLesson {
        const lines = extractLines(lessonCell);
        if (!lines.length) {
            return null;
        }

        const entries = this.parseEntries(lines);
        if (!entries.length && config.parser.v2?.strict) {
            throw new Error('unable to parse lesson entries');
        }
        if (!entries.length) {
            return null;
        }

        const cabinets = extractLines(cabinetCell).map((value) => {
            return this.removeDashes(value) ?? null;
        });

        this.applyCabinets(entries, cabinets);

        if (entries.length === 1 && entries[0].subgroup === undefined) {
            return entries[0];
        }

        return entries;
    }

    protected parseEntries(lines: string[]): GroupLessonExplain[] {
        const entries: GroupLessonExplain[] = [];
        let current: Partial<GroupLessonExplain> = {};

        const pushCurrent = () => {
            if (!current.lesson) {
                current = {};
                return;
            }

            entries.push({
                subgroup: current.subgroup,
                lesson: current.lesson,
                type: current.type ?? null,
                teacher: current.teacher ?? null,
                cabinet: null,
                comment: null
            });

            current = {};
        };

        for (const raw of lines) {
            const combined = this.parseCombinedLine(raw);
            if (combined) {
                pushCurrent();
                entries.push(combined);
                continue;
            }

            const type = this.parseType(raw);
            if (type && current.lesson) {
                current.type = type;
                continue;
            }

            if (!current.lesson) {
                const parsed = this.parseLessonLine(raw);
                if (parsed) {
                    current.lesson = parsed.lesson;
                    current.subgroup = parsed.subgroup;
                }
                continue;
            }

            if (!current.teacher) {
                current.teacher = raw;
                continue;
            }

            if (isTeacherLine(raw)) {
                current.teacher = `${current.teacher}, ${raw}`;
                continue;
            }

            pushCurrent();
            const parsed = this.parseLessonLine(raw);
            if (parsed) {
                current.lesson = parsed.lesson;
                current.subgroup = parsed.subgroup;
            }
        }

        pushCurrent();

        return entries;
    }

    protected getCell(row: GridCell[], rowIndex: number, colIndex: number): HTMLTableCellElement | null {
        const cell = row[colIndex];
        if (!cell) {
            return null;
        }

        if (cell.row !== rowIndex) {
            return null;
        }

        return cell.cell;
    }

    protected parseCombinedLine(line: string): GroupLessonExplain | null {
        const match = line.match(/^(?:(\d+)\s*[.)]\s*)?(.+?)\s*\(([^)]+)\)\s*(.+)?$/);
        if (!match) {
            return null;
        }

        const subgroup = match[1] ? Number(match[1]) : undefined;
        const lesson = getShortSubjectName(match[2].trim());
        const type = match[3].trim();
        const teacher = cleanText(match[4]) ?? null;

        return {
            subgroup,
            lesson,
            type: type || null,
            teacher,
            cabinet: null,
            comment: null
        };
    }

    protected parseType(line: string): string | null {
        const match = line.match(/^\((.+)\)$/);
        return match ? match[1].trim() : null;
    }

    protected parseLessonLine(line: string): { lesson: string, subgroup?: number } | null {
        const match = line.match(/^(?:(\d+)\s*[.)]\s*)?(.+)$/);
        if (!match) {
            return null;
        }

        const subgroup = match[1] ? Number(match[1]) : undefined;
        const lesson = getShortSubjectName(match[2].trim());

        return {
            lesson,
            subgroup
        };
    }

    protected applyCabinets(entries: GroupLessonExplain[], cabinets: (string | null)[]) {
        if (entries.length === 0) {
            return;
        }

        const normalized = cabinets.length ? cabinets : [null];
        const hasAny = normalized.some((value) => value);
        if (!hasAny) {
            return;
        }

        if (normalized.length === 1) {
            for (const entry of entries) {
                entry.cabinet = normalized[0] ?? null;
            }
            return;
        }

        const hasSubgroup = entries.some((entry) => entry.subgroup !== undefined);
        if (hasSubgroup) {
            for (let i = 0; i < entries.length; i++) {
                const entry = entries[i];
                if (entry.subgroup !== undefined && normalized[entry.subgroup - 1] !== undefined) {
                    entry.cabinet = normalized[entry.subgroup - 1] ?? null;
                } else {
                    entry.cabinet = normalized[Math.min(i, normalized.length - 1)] ?? null;
                }
            }
            return;
        }

        for (let i = 0; i < entries.length; i++) {
            entries[i].cabinet = normalized[i] ?? normalized[normalized.length - 1] ?? null;
        }
    }

    private postProcessDay(day: GroupDay) {
        for (let i = 0; i <= day.lessons.length; i++) {
            let lesson: GroupLesson = day.lessons[i];
            if (!lesson) continue;

            if (!Array.isArray(lesson)) {
                lesson = [lesson];
            }

            if (lesson.every((_) => _.type === 'ф-в' && _.comment == null)) {
                let simmilarIndex: number | null = null;

                let firstNotNull = false;
                for (let j: number = day.lessons.length; j > i; j--) {
                    let fLesson: GroupLesson = day.lessons[j];

                    if (firstNotNull && !fLesson) break;
                    if (!fLesson) continue;
                    firstNotNull = true;

                    if (!Array.isArray(fLesson)) {
                        fLesson = [fLesson];
                    }

                    if (lesson.length !== fLesson.length) continue;

                    for (let k = 0; k < lesson.length; k++) {
                        if (
                            lesson[k].type === fLesson[k].type &&
                            lesson[k].lesson === fLesson[k].lesson &&
                            lesson[k].teacher === fLesson[k].teacher
                        ) {
                            simmilarIndex = j;
                            break;
                        }
                    }

                    if (simmilarIndex !== null) {
                        break;
                    }
                }

                if (simmilarIndex !== null) {
                    lesson.forEach((_) => _.comment = '2 часа');
                    day.lessons[simmilarIndex] = null;
                }
            }
        }

        this.clearEndingNull(day.lessons);
    }
}

File: internal/parser/v2/group.go
package v2

import (
	"fmt"
	"regexp"
	"strings"
	"time"

	"github.com/PuerkitoBio/goquery"
	"github.com/blindmaster24/MgkeTimetableBot/internal/model"
)

type GroupParserV2 struct {
	doc *goquery.Document
}

func NewGroupParserV2(doc *goquery.Document) *GroupParserV2 {
	return &GroupParserV2{doc: doc}
}

func (p *GroupParserV2) ContentHash() string {
	text := p.doc.Text()
	h := [32]byte{}
	copy(h[:], []byte(text))
	return fmt.Sprintf("%x", h)
}

func (p *GroupParserV2) Run() (model.Groups, error) {
	groups := make(model.Groups)
	tables := p.findTables()

	for _, table := range tables {
		label := p.findHeading(table, "группа")
		if label == "" {
			continue
		}

		parsed := p.parseTable(table, label)
		if parsed == nil {
			continue
		}

		existing := groups[parsed.GroupNumber]
		if existing != nil {
			existing.Days = mergeGroupDays(parsed.Days, existing.Days)
			existing.Group = parsed.Group
		} else {
			groups[parsed.GroupNumber] = &model.Group{
				Group: parsed.Group,
				Days:  parsed.Days,
			}
		}
	}

	clearSundays(groups)

	for _, g := range groups {
		for i := range g.Days {
			postProcessGroupDay(&g.Days[i])
		}
	}

	return groups, nil
}

type parsedGroupTable struct {
	Group       string
	GroupNumber string
	Days        []model.GroupDay
}

func (p *GroupParserV2) parseTable(table *goquery.Selection, label string) *parsedGroupTable {
	group := parseGroupLabel(label)
	groupNumber := extractGroupNumber(group)
	if group == "" || groupNumber == "" {
		return nil
	}

	grid := BuildTableGrid(table)
	headerRowIndex := FindHeaderRowIndex(grid, 5, 5)
	if headerRowIndex < 0 {
		return nil
	}
	ranges := GetDayRangesFromGrid(grid, headerRowIndex)
	if len(ranges) < 5 {
		return nil
	}

	days := make([]model.GroupDay, len(ranges))
	for i, r := range ranges {
		days[i] = model.GroupDay{Day: r.Day, Lessons: []model.GroupLesson{}}
	}

	for rowIndex := headerRowIndex + 1; rowIndex < grid.Height; rowIndex++ {
		row := grid.Grid[rowIndex]
		if row == nil || len(row) == 0 {
			continue
		}

		numberCell := row[0]
		if numberCell == nil || numberCell.Row != rowIndex {
			continue
		}

		lessonNumber := parseLessonNumber(numberCell.Cell.Text())
		if lessonNumber == 0 {
			continue
		}

		for i, r := range ranges {
			lessonCell := getCellSafe(row, rowIndex, r.Start)
			var cabinetCell *goquery.Selection
			if r.Span > 1 {
				cabinetCell = getCellSafe(row, rowIndex, r.Start+1)
			}

			lesson := parseGroupLesson(lessonCell, cabinetCell)
			assignGroupLesson(&days[i], lessonNumber, lesson)
		}
	}

	return &parsedGroupTable{
		Group:       group,
		GroupNumber: groupNumber,
		Days:        days,
	}
}

func (p *GroupParserV2) findTables() []*goquery.Selection {
	var tables []*goquery.Selection
	p.doc.Find("table").Each(func(_ int, table *goquery.Selection) {
		grid := BuildTableGrid(table)
		hri := FindHeaderRowIndex(grid, 5, 5)
		if hri < 0 {
			return
		}
		ranges := GetDayRangesFromGrid(grid, hri)
		if len(ranges) < 5 {
			return
		}
		tables = append(tables, table)
	})
	return tables
}

func (p *GroupParserV2) findHeading(table *goquery.Selection, keyword string) string {
	prev := table.Prev()
	depth := 0
	for prev.Length() > 0 && depth < 12 {
		tag := goquery.NodeName(prev)
		if tag == "h1" || tag == "h2" || tag == "h3" || tag == "h4" {
			text := cleanText(prev.Text())
			if text != "" && strings.Contains(strings.ToLower(text), keyword) {
				return text
			}
		}
		if tag == "table" {
			break
		}
		prev = prev.Prev()
		depth++
	}
	return ""
}

func parseGroupLabel(label string) string {
	text := cleanText(label)
	if text == "" || !strings.Contains(strings.ToLower(text), "группа") {
		return ""
	}
	parts := strings.SplitN(text, "-", 2)
	if len(parts) < 2 {
		return ""
	}
	return strings.TrimSpace(parts[1])
}

func extractGroupNumber(group string) string {
	re := regexp.MustCompile(`\d+`)
	return re.FindString(group)
}

func parseGroupLesson(lessonCell, cabinetCell *goquery.Selection) model.GroupLesson {
	lines := extractLines(lessonCell)
	if len(lines) == 0 {
		return nil
	}

	entries := parseGroupEntries(lines)
	if len(entries) == 0 {
		return nil
	}

	cabinets := extractLines(cabinetCell)
	applyCabinets(entries, cabinets)

	if len(entries) == 1 && entries[0].Subgroup == nil {
		return entries[0]
	}
	return entries
}

func parseGroupEntries(lines []string) []*model.GroupLessonExplain {
	var entries []*model.GroupLessonExplain
	var sub *int
	var lesson, typ, teacher string

	push := func() {
		if lesson == "" {
			sub = nil
			lesson = ""
			typ = ""
			teacher = ""
			return
		}
		e := &model.GroupLessonExplain{
			Subgroup: sub,
			Lesson:   lesson,
		}
		if typ != "" {
			e.Type = &typ
		}
		if teacher != "" {
			e.Teacher = &teacher
		}
		entries = append(entries, e)
		sub = nil
		lesson = ""
		typ = ""
		teacher = ""
	}

	for _, raw := range lines {
		if c := parseCombinedLine(raw); c != nil {
			push()
			entries = append(entries, c)
			continue
		}
		if t := parseTypeLine(raw); t != nil && lesson != "" {
			typ = *t
			continue
		}
		if lesson == "" {
			if p := parseLessonLine(raw); p != nil {
				lesson = p.Lesson
				sub = p.Subgroup
			}
			continue
		}
		if teacher == "" {
			teacher = raw
			continue
		}
		if isTeacherLine(raw) {
			teacher = teacher + ", " + raw
			continue
		}
		push()
		if p := parseLessonLine(raw); p != nil {
			lesson = p.Lesson
			sub = p.Subgroup
		}
	}
	push()
	return entries
}

var combinedRe = regexp.MustCompile(`^(?:(\d+)\s*[.)]\s*)?(.+?)\s*\(([^)]+)\)\s*(.+)?$`)

func parseCombinedLine(line string) *model.GroupLessonExplain {
	match := combinedRe.FindStringSubmatch(line)
	if match == nil {
		return nil
	}
	var sub *int
	if match[1] != "" {
		v := 0
		fmt.Sscanf(match[1], "%d", &v)
		sub = &v
	}
	l := strings.TrimSpace(match[2])
	t := strings.TrimSpace(match[3])
	te := cleanText(match[4])
	return &model.GroupLessonExplain{Subgroup: sub, Lesson: l, Type: &t, Teacher: &te}
}

func parseTypeLine(line string) *string {
	s := strings.TrimSpace(line)
	if len(s) > 2 && s[0] == '(' && s[len(s)-1] == ')' {
		v := s[1 : len(s)-1]
		return &v
	}
	return nil
}

var lessonLineRe = regexp.MustCompile(`^(?:(\d+)\s*[.)]\s*)?(.+)$`)

type parsedLessonLine struct {
	Lesson   string
	Subgroup *int
}

func parseLessonLine(line string) *parsedLessonLine {
	match := lessonLineRe.FindStringSubmatch(line)
	if match == nil {
		return nil
	}
	var sub *int
	if match[1] != "" {
		v := 0
		fmt.Sscanf(match[1], "%d", &v)
		sub = &v
	}
	return &parsedLessonLine{Lesson: strings.TrimSpace(match[2]), Subgroup: sub}
}

func applyCabinets(entries []*model.GroupLessonExplain, cabinets []string) {
	if len(entries) == 0 {
		return
	}
	normalized := make([]string, len(cabinets))
	copy(normalized, cabinets)
	if len(normalized) == 0 {
		normalized = []string{""}
	}
	hasAny := false
	for _, v := range normalized {
		if v != "" && v != "-" {
			hasAny = true
			break
		}
	}
	if !hasAny {
		return
	}
	if len(normalized) == 1 {
		val := normalized[0]
		for _, e := range entries {
			e.Cabinet = &val
		}
		return
	}
	hasSub := false
	for _, e := range entries {
		if e.Subgroup != nil {
			hasSub = true
			break
		}
	}
	if hasSub {
		for i, e := range entries {
			idx := i
			if e.Subgroup != nil && *e.Subgroup-1 >= 0 && *e.Subgroup-1 < len(normalized) {
				idx = *e.Subgroup - 1
			} else if idx >= len(normalized) {
				idx = len(normalized) - 1
			}
			val := normalized[idx]
			e.Cabinet = &val
		}
		return
	}
	for i, e := range entries {
		idx := i
		if idx >= len(normalized) {
			idx = len(normalized) - 1
		}
		val := normalized[idx]
		e.Cabinet = &val
	}
}

func assignGroupLesson(day *model.GroupDay, lessonNumber int, lesson model.GroupLesson) {
	index := lessonNumber - 1
	for len(day.Lessons) < index {
		day.Lessons = append(day.Lessons, nil)
	}
	if len(day.Lessons) == index {
		day.Lessons = append(day.Lessons, lesson)
	} else {
		day.Lessons[index] = lesson
	}
}

func getCellSafe(row []*GridCell, rowIndex, colIndex int) *goquery.Selection {
	if colIndex < 0 || colIndex >= len(row) || row[colIndex] == nil {
		return nil
	}
	cell := row[colIndex]
	if cell.Row != rowIndex {
		return nil
	}
	return cell.Cell
}

func postProcessGroupDay(day *model.GroupDay) {
	for i := 0; i < len(day.Lessons); i++ {
		lesson := day.Lessons[i]
		if lesson == nil {
			continue
		}
		arr := model.AsArray(lesson)
		if arr == nil {
			if s := model.AsSingle(lesson); s != nil {
				arr = []*model.GroupLessonExplain{s}
			}
		}
		if arr == nil {
			continue
		}
		allPhv := true
		for _, e := range arr {
			if e.Type == nil || *e.Type != "ф-в" || e.Comment != nil {
				allPhv = false
				break
			}
		}
		if !allPhv {
			continue
		}
		similarIdx := -1
		firstNotNull := false
		for j := len(day.Lessons) - 1; j > i; j-- {
			fl := day.Lessons[j]
			if firstNotNull && fl == nil {
				break
			}
			if fl == nil {
				continue
			}
			firstNotNull = true
			fArr := model.AsArray(fl)
			if fArr == nil {
				if fs := model.AsSingle(fl); fs != nil {
					fArr = []*model.GroupLessonExplain{fs}
				}
			}
			if fArr == nil || len(fArr) != len(arr) {
				continue
			}
			match := true
			for k := 0; k < len(arr); k++ {
				if arr[k].Type != fArr[k].Type || arr[k].Lesson != fArr[k].Lesson || arr[k].Teacher != fArr[k].Teacher {
					match = false
					break
				}
			}
			if match {
				similarIdx = j
				break
			}
		}
		if similarIdx >= 0 {
			comment := "2 часа"
			for _, e := range arr {
				e.Comment = &comment
			}
			day.Lessons[similarIdx] = nil
		}
	}
	for len(day.Lessons) > 0 && day.Lessons[len(day.Lessons)-1] == nil {
		day.Lessons = day.Lessons[:len(day.Lessons)-1]
	}
}

func mergeGroupDays(newDays, oldDays []model.GroupDay) []model.GroupDay {
	days := make(map[string]*model.GroupDay)
	for i := range oldDays {
		days[oldDays[i].Day] = &oldDays[i]
	}
	for i := range newDays {
		days[newDays[i].Day] = &newDays[i]
	}
	var result []model.GroupDay
	for _, d := range days {
		result = append(result, *d)
	}
	return result
}

func clearSundays(groups model.Groups) {
	for _, g := range groups {
		var filtered []model.GroupDay
		for _, d := range g.Days {
			t, err := time.Parse("02.01.2006", d.Day)
			if err == nil && t.Weekday() != time.Sunday {
				filtered = append(filtered, d)
			}
		}
		g.Days = filtered
	}
}
